src/psbt.c

```diff
@@ -287,303 +332,301 @@ int wally_unknowns_map_free(struct wally_unknowns_map *unknowns)
 }
 
-int wally_psbt_input_set_final_script_sig(
-    struct wally_psbt_input *input,
-    unsigned char *final_script_sig,
-    size_t final_script_sig_len)
+int wally_psbt_input_set_keypaths(struct wally_psbt_input *input,
+                                  const struct wally_keypath_map *keypaths)
 {
-    unsigned char *result_final_script_sig;
+    return input ? replace_keypaths(keypaths, &input->keypaths) : WALLY_EINVAL;
+}
 
-    if (!clone_bytes(&result_final_script_sig, final_script_sig, final_script_sig_len)) {
-        return WALLY_ENOMEM;
-    }
-    wally_free(input->final_script_sig);
-    input->final_script_sig = result_final_script_sig;
-    input->final_script_sig_len = final_script_sig_len;
-    return WALLY_OK;
+int wally_psbt_input_set_partial_sigs(struct wally_psbt_input *input,
+                                      const struct wally_partial_sigs_map *partial_sigs)
+{
+    return input ? replace_partial_sigs(partial_sigs, &input->partial_sigs) : WALLY_EINVAL;
 }
 
-int wally_psbt_input_set_final_witness(
-    struct wally_psbt_input *input,
-    struct wally_tx_witness_stack *final_witness)
+int wally_psbt_input_set_unknowns(struct wally_psbt_input *input,
+                                  const struct wally_unknowns_map *unknowns)
 {
-    struct wally_tx_witness_stack *result_final_witness;
+    return input ? replace_unknowns(unknowns, &input->unknowns) : WALLY_EINVAL;
+}
 
-    if (!(result_final_witness = clone_witness(final_witness))) {
-        return WALLY_ENOMEM;
-    }
-    wally_tx_witness_stack_free(input->final_witness);
-    input->final_witness = result_final_witness;
+int wally_psbt_input_set_sighash_type(struct wally_psbt_input *input, uint32_t sighash_type)
+{
+    if (!input)
+        return WALLY_EINVAL;
+    input->sighash_type = sighash_type;
     return WALLY_OK;
 }
 
-int wally_psbt_input_set_keypaths(
-    struct wally_psbt_input *input,
-    struct wally_keypath_map *keypaths)
+#ifdef BUILD_ELEMENTS
+int wally_psbt_elements_input_set_value(struct wally_psbt_input *input, uint64_t value)
 {
-    struct wally_keypath_map *result_keypaths;
+    input->value = value;
+    input->has_value = 1u;
+    return WALLY_OK;
+}
 
-    if (!(result_keypaths = clone_keypath_map(keypaths))) {
-        return WALLY_ENOMEM;
-    }
-    wally_keypath_map_free(input->keypaths);
-    input->keypaths = result_keypaths;
+int wally_psbt_elements_input_clear_value(struct wally_psbt_input *input)
+{
+    input->value = 0;
+    input->has_value = 0;
     return WALLY_OK;
 }
 
-int wally_psbt_input_set_partial_sigs(
-    struct wally_psbt_input *input,
-    struct wally_partial_sigs_map *partial_sigs)
+int wally_psbt_elements_input_set_value_blinder(struct wally_psbt_input *input,
+                                                const unsigned char *value_blinder,
+                                                size_t value_blinder_len)
 {
-    struct wally_partial_sigs_map *result_partial_sigs;
+    if (!input)
```

Comment:
I think we should check (here and below) the length of the value set against the values in https://github.com/ElementsProject/libwally-core/blob/master/include/wally_elements.h#L12. This would be consistent with what is done in `wally_psbt_elements_output_set_blinding_pubkey`

Reply:
Yes, agreed. I would have added this but at present the naming of things leaves me a little confused as to what these members actually are. Is "asset_blinder" _actually_ an asset blinding factor? If so it  should be called abf and passed with abf_len so SWIG picks it up automatically and the docs are easier to follow and maintain.

In general the code prior to this patch series does no testing of values, for example pubkey validity and preventing dupes in KV maps. I think tightening up the validation will happen iteratively after this patch series is applied.

I will add a FIXME to check these values until I can confirm what they actually represent by following through to where they are used. These setters will probably be replaced with macro versions at some point as well.

Reply:
FIXME added.